File: infrastructures/ddd/csharp/src/Domain/Models/BaseModel.cs
using System;

namespace OddlyDdd.Domain.Models
{
    /// <summary>
    /// Base class for all Business Model Objects (BMOs).
    /// BMOs contain business logic and invariants.
    /// CRITICAL: BMOs MUST NOT have database/ORM attributes.
    /// They live in /Domain/Models/ and have Model or BMO suffix.
    /// Separation from persistence layer is MANDATORY.
    /// </summary>
    public abstract class BaseModel
    {
        protected string _id = string.Empty;
        protected DateTime _createdAt;
        protected DateTime _updatedAt;

        public string Id
        {
            get => _id;
            protected set => _id = value;
        }

        public DateTime CreatedAt
        {
            get => _createdAt;
            protected set => _createdAt = value;
        }

        public DateTime UpdatedAt
        {
            get => _updatedAt;
            protected set => _updatedAt = value;
        }

        protected BaseModel()
        {
            _id = Guid.NewGuid().ToString();
            _createdAt = DateTime.UtcNow;
            _updatedAt = DateTime.UtcNow;
        }

        /// <summary>
        /// Validates the business invariants of the model.
        /// Override this method to add specific validation logic.
        /// </summary>
        public virtual void Validate()
        {
            if (string.IsNullOrWhiteSpace(_id))
            {
                throw new InvalidOperationException("Model ID cannot be empty");
            }
        }
    }
}

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/domain/models/infra/BaseModel.java
package com.oddly.ddd.domain.models.infra;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Base class for all Business Model Objects (BMOs).
 * BMOs contain business logic and invariants.
 * CRITICAL: BMOs MUST NOT have database/ORM attributes.
 * They live in /domain/models/ and have Model or BMO suffix.
 * Separation from persistence layer is MANDATORY.
 */
public abstract class BaseModel {
    protected String m_id;
    protected LocalDateTime m_createdAt;
    protected LocalDateTime m_updatedAt;

    protected BaseModel() {
        this.m_id = UUID.randomUUID().toString();
        this.m_createdAt = LocalDateTime.now();
        this.m_updatedAt = LocalDateTime.now();
    }

    protected BaseModel(String p_id) {
        this.m_id = p_id;
        this.m_createdAt = LocalDateTime.now();
        this.m_updatedAt = LocalDateTime.now();
    }

    public String getId() {
        return m_id;
    }

    protected void setId(String p_id) {
        this.m_id = p_id;
    }

    public LocalDateTime getCreatedAt() {
        return m_createdAt;
    }

    protected void setCreatedAt(LocalDateTime p_createdAt) {
        this.m_createdAt = p_createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return m_updatedAt;
    }

    protected void setUpdatedAt(LocalDateTime p_updatedAt) {
        this.m_updatedAt = p_updatedAt;
    }

    /**
     * Updates the updatedAt timestamp to current time.
     * Should be called when the model is modified.
     */
    protected void touch() {
        this.m_updatedAt = LocalDateTime.now();
    }

    /**
     * Validates the business invariants of the model.
     * Override this method to add specific validation logic.
     * @throws IllegalStateException if validation fails
     */
    public void validate() {
        if (m_id == null || m_id.isBlank()) {
            throw new IllegalStateException("Model ID cannot be empty");
        }
    }
}
